public/pages/Administration/components/webhook/WebhookForm.tsx

```diff
@@ -0,0 +1,286 @@
+import "./WebhookForm.scss"
+
+import React, { useState } from "react"
+import { Button, Field, Form, Input, Loader, Message, Select, SelectOption, TextArea, Toggle } from "@fider/components"
+import { actions, Failure } from "@fider/services"
+import { HStack, VStack } from "@fider/components/layout"
+import { Webhook, WebhookPreviewResult, WebhookStatus, WebhookType } from "@fider/models"
+import { HoverInfo } from "@fider/components/common/HoverInfo"
+import { WebhookTemplateInfo } from "@fider/pages/Administration/components/webhook/WebhookTemplateInfo"
+
+interface WebhookFormProps {
+  webhook?: Webhook
+  onSave: (data: WebhookFormState) => Promise<Failure | undefined>
+  onCancel: () => void
+}
+
+export interface WebhookFormState extends Omit<Webhook, "id"> {
+  error?: Failure
+  typing?: NodeJS.Timeout
+  preview?: WebhookPreviewResult | null
+  isModalOpen: boolean
+}
+
+interface HttpHeaderProps {
+  header?: string
+  value?: string
+  onEdit: (header: string, value: string) => void
+  onRemove?: (header: string) => void
+  allHeaders?: string[]
+}
+
+const HttpHeader = (props: HttpHeaderProps) => {
+  const [header, setHeader] = useState(props.header || "")
+  const [value, setValue] = useState(props.value || "")
+  const [editing, setEditing] = useState(props.onRemove === undefined)
+
+  const duplicate = props.allHeaders && props.allHeaders.includes(header)
+
+  const suffix = editing ? (
+    props.onRemove === undefined ? (
+      <Button
+        variant="primary"
+        onClick={() => {
+          props.onEdit(header, value)
+          setHeader("")
+          setValue("")
+        }}
+        disabled={duplicate || header.length === 0 || value.length === 0}
+      >
+        Add
+      </Button>
+    ) : (
+      <Button
+        variant="primary"
+        onClick={() => {
+          setEditing(false)
+          props.onEdit(header, value)
+        }}
+        disabled={value.length === 0}
+      >
+        Save
+      </Button>
+    )
+  ) : (
+    <>
+      <Button variant="secondary" onClick={() => setEditing(true)}>
+        Edit
+      </Button>
+      <Button variant="danger" onClick={() => props.onRemove && props.onRemove(header)}>
+        Remove
+      </Button>
+    </>
+  )
+
+  return (
+    <HStack justify="full" spacing={4}>
+      <Input
+        className="mb-0"
+        field={`header-${header}`}
+        value={header}
+        onChange={setHeader}
+        placeholder="Header"
+        disabled={props.onRemove !== undefined}
+        suffix={duplicate ? "Duplicate" : undefined}
+      />
+      <Input field={`value-${header}`} value={value} onChange={setValue} placeholder="Value" disabled={!editing} suffix={suffix} />
+    </HStack>
+  )
+}
+
+export class WebhookForm extends React.Component<WebhookFormProps, WebhookFormState> {
```

Comment:
I'm slowly migration all older class components into functions+hook. Do you mind converting this one so we don't add new tech debt?

Reply:
Yep, no problem I understand. Good to use those (no so) new React features